### PR TITLE
[Improvement] Draw markers faster

### PR DESCRIPTION
**Pull request by ahuarte47**
The code takes advantage of points already created to paint the geometries to draw the markers.

### CODE REVIEW

**Comment by m-kuhn** [comment]
Looks good to me from a logic perspective.

Could you add some more linebreaks to increase readability (after if, else, for...)?

What do you think about using the `Q_FOREACH` macro instead of for loops and pointer arithmetic / random access by using `at( i )`? Are you afraid of a performance impact?

**Comment by ahuarte47** [comment]
I changed it, thanks @m-kuhn ! 

About the Q_FOREACH macro, IMHO the performance impact should be minimum compared to time used by the full rendering task.

File: src/core/symbology-ng/qgssymbolv2.cpp [hunk 1]
```diff
@@ -785,6 +798,16 @@ void QgsSymbolV2::renderFeature( const QgsFeature& feature, QgsRenderContext& co
       }
       _getPolygon( pts, holes, context, QgsConstWkbPtr( segmentizedGeometry->asWkb(), segmentizedGeometry->wkbSize() ), !tileMapRendering && clipFeaturesToExtent() );
       static_cast<QgsFillSymbolV2*>( this )->renderPolygon( pts, ( !holes.isEmpty() ? &holes : nullptr ), &feature, context, layer, selected );
+
+      if ( drawVertexMarker )
+      {
+        markers = pts;
+
+        Q_FOREACH ( QPolygonF hole, holes )
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.cpp]: You'll want to use `const QPolygonF&` here, will be a bit faster

Comment (ahuarte47) [inline src/core/symbology-ng/qgssymbolv2.cpp]: Changed, thanks a lot!

**Comment by m-kuhn** [comment]
Waiting for travis to be ok and then it's ready for a public testdrive.

Thanks @ahuarte47

**Comment by ahuarte47** [comment]
Ummm, Travis throws 9 build errors, There is other reasons ?

File: src/core/symbology-ng/qgssymbolv2.cpp [hunk 2]
```diff
@@ -903,27 +965,9 @@ void QgsSymbolV2::renderFeature( const QgsFeature& feature, QgsRenderContext& co
...
   {
-    const QgsCoordinateTransform* ct = context.coordinateTransform();
-    const QgsMapToPixel& mtp = context.mapToPixel();
-
-    QgsPointV2 vertexPoint;
-    QgsVertexId vertexId;
-    double x, y, z;
-    QPointF mapPoint;
-    while ( geom->geometry()->nextVertex( vertexId, vertexPoint ) )
+    Q_FOREACH ( const QPointF& marker, markers )
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.cpp]: @ahuarte47 Note that for this one, you actually want a copy rather then the reference. QPointF is small and it's quicker to copy than resolve a reference. So this should be `Q_FOREACH( QPointF marker, markers )` (this rule basically only applies to small inbuilt types like int, double, etc AND some very small Qt objects like QPoint, QPointF, QSize and QSizeF)

**Comment by nyalldawson** [comment]
@ahuarte47 really nice! This also has the benefit of using the clipped geometry to draw markers, instead of drawing a bunch of markers which aren't visible in the current map extent.

Those build failures are unrelated - it's a clang error which we've seen once before. Let me test something and I'll get back to you on this.

**Comment by ahuarte47** [comment]
Hy Nyall, Thanks alot for all support!, all days I learn from your great knowledge!

**Comment by nyalldawson** [comment]
@ahuarte47 can you try rebasing against master now?

**Comment by ahuarte47** [comment]
Sure!

**Comment by m-kuhn** [comment]
Thanks a lot!

Will you be looking into tinywkb next? ;)

**Comment by ahuarte47** [comment]
Thanks a lot Mathias and Nyall for your support!

Now, I am developing a new pull about [simplification](https://github.com/qgis/QGIS/pull/2744#issuecomment-190717940).

But I gets a big performance issue using QgsGeos in [QgsGeometry](https://github.com/qgis/QGIS/pull/2859)

**Pull request merged**